org.eclipse.xtext.xtext.ui.examples/projects/homeautomation/org.eclipse.xtext.example.homeautomation/src/org/eclipse/xtext/example/homeautomation/validation/RuleEngineValidator.java

```diff
@@ -0,0 +1,109 @@
+/**
+ * Copyright (c) 2015, 2019 itemis AG (http://www.itemis.eu) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.xtext.example.homeautomation.validation;
+
+import static com.google.common.base.Objects.equal;
+
+import java.util.HashSet;
+
+import javax.inject.Inject;
+
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.xtext.EcoreUtil2;
+import org.eclipse.xtext.common.types.JvmOperation;
+import org.eclipse.xtext.example.homeautomation.jvmmodel.RuleEngineJvmModelInferrer;
+import org.eclipse.xtext.example.homeautomation.ruleEngine.Declaration;
+import org.eclipse.xtext.example.homeautomation.ruleEngine.Device;
+import org.eclipse.xtext.example.homeautomation.ruleEngine.Model;
+import org.eclipse.xtext.example.homeautomation.ruleEngine.Rule;
+import org.eclipse.xtext.example.homeautomation.ruleEngine.RuleEnginePackage;
+import org.eclipse.xtext.validation.Check;
+import org.eclipse.xtext.xbase.XAbstractFeatureCall;
+import org.eclipse.xtext.xbase.XExpression;
+import org.eclipse.xtext.xbase.XFeatureCall;
+import org.eclipse.xtext.xbase.XbasePackage;
+import org.eclipse.xtext.xbase.jvmmodel.IJvmModelAssociations;
+import org.eclipse.xtext.xbase.lib.CollectionLiterals;
+import org.eclipse.xtext.xbase.lib.StringExtensions;
+
+/**
+ * This class contains custom validation rules.
+ * 
+ * See
+ * https://www.eclipse.org/Xtext/documentation/303_runtime_concepts.html#validation
+ */
+public class RuleEngineValidator extends AbstractRuleEngineValidator {
+	@Inject
+	private IJvmModelAssociations jvmModelAssociations;
+
+	@Check
+	public void checkUniqueDeclarations(Model model) {
+		HashSet<String> deviceNames = CollectionLiterals.<String>newHashSet();
```

Comment:
I'd prefer a plain constructor call. Local vars could be `Set` instead of `HashSet`